modules/core/ui/views/src/Plugin/views/argument/EntityTaxonomyTermReferenceArgument.php

```diff
@@ -0,0 +1,176 @@
+    EntityFieldManagerInterface $entity_field_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $term_storage);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityTypeBundleInfo = $entity_bundle_info;
+    $this->entityFieldManager = $entity_field_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static($configuration, $plugin_id, $plugin_definition,
+      $container->get('entity_type.manager')->getStorage('taxonomy_term'),
+      $container->get('entity_type.manager'),
+      $container->get('entity_type.bundle.info'),
+      $container->get('entity_field.manager'));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function query($group_by = FALSE) {
+    $term_id = $this->argument;
+    $entity_bundle = $this->view->args[1] ?: 'all';
+
+    if (empty($term_id)) {
+      return;
+    }
+
+    $term = $this->termStorage->load($term_id);
+
+    // This is a value like 'asset' or 'log'.
+    $base_entity_type = $this->view->getBaseEntityType()->id();
+
+    $entity_storage = $this->entityTypeManager->getStorage($base_entity_type);
+
+    if (!($entity_storage instanceof SqlContentEntityStorage)) {
+      return;
+    }
+
+    $entity_data_table = $entity_storage->getDataTable();
+
+    $entity_table_mapping = $entity_storage->getTableMapping();
+
+    $conditions = [];
+
+    // Keep track of which field tables we've already joined with since some
+    // assets share the same field e.g. plant and seed assets.
+    $already_joined_term_field_tables = [];
+
+    $bundles = $this->entityTypeBundleInfo->getBundleInfo($base_entity_type);
+    foreach (array_keys($bundles) as $type) {
+      if ($entity_bundle == 'all' || $type == $entity_bundle) {
+        $field_definitions = $this->entityFieldManager->getFieldDefinitions($base_entity_type, $type);
+
+        foreach ($field_definitions as $field_id => $field_definition) {
+          if ($field_definition->getType() == "entity_reference" && $field_definition->getSetting('target_type') == "taxonomy_term") {
+            $handler_settings = $field_definition->getSetting('handler_settings') ?? [];
+
+            if (in_array($term->bundle(), $handler_settings['target_bundles'] ?? [])) {
+
+              $field_table_name = $entity_table_mapping->getFieldTableName($field_id);
+
+              if (array_key_exists($field_table_name, $already_joined_term_field_tables)) {
+                continue;
+              }
+
+              $column_names = $entity_table_mapping->getColumnNames($field_id);
+
+              $target_id_column_name = $column_names['target_id'];
+
+              // Join the taxonomy reference field table with the entity.
+              /** @var \Drupal\views\Plugin\views\join\JoinPluginBase $join */
+              $join = Views::pluginManager('join')->createInstance('standard', [
```

Comment:
I think I need a bit more clarification here...

What I thought I understood from the conversation on the call was that there might be some way to leverage the existing Views configuration that defines the entity/field table relationships to avoid needing to specify a join at the `EntityQuery` layer - as I'm doing currently.

However, in looking at both that blog post and the `hook_views_data` documentation, I think I'm seeing that code doing exactly the same thing. i.e. the code I might add in `hook_views_data` (or `hook_views_data_alter`) would still need to know about (or look up) the relevant table/column names.

If that's true, then it would just be a question of where it is cleanest to put the code that adds the join.

In `hook_views_data` I'd presumably need to have some conditional logic based on the view display (e.g. `page_term`) to add the relevant joins and separately I'd still need code in the views argument plugin to and the `WHERE` clauses.

Assuming my understanding above is correct, it seems cleaner to keep all the join/query code in the argument plugin since then the logic is all in one place and views takes care of only invoking the plugin for displays which use the argument.

Am I missing something?

Reply:
Yea I think the main question was whether or not we could separate the "definition of the relationship" from the "adding of the JOIN", if that makes sense. But I don't know if it warrants reworking this right now.

`hook_views_data()` (and `hook_views_data_alter()`) are typically where you describe the relationships between tables. These hooks only run when the Views data cache is rebuilt (which happens less frequently than content caches being rebuilt, as would be the case with the logic in the Views argument). So *maybe* there would be a slight performance benefit to describing the relationships in there? It would also mean that those relationships are available for use in other code, potentially.

As for actually *adding* the JOIN, that would still need to happen in the Views argument class regardless, once you know which tables you need. But if the relationship is already described to Views in `hook_views_data()` then adding the join itself would be simpler (just a single method call I believe).

So, the two possible advantages to using `hook_views_data()`:

1. Performance improvement (but maybe only if it avoids running expensive field data lookups in the Views argument handler)
2. Reusable relationships (if someone wanted to build a custom View that used one of these relationships in the future, for instance)

Full disclosure: I'm just describing the high-level thinking here, but haven't fully thought through the specific details of this case, so it could be that those too advantages are not even applicable in this case.

In either case, I'm fine with just leaving the logic in one place for now. We can always consider moving it later, if there were a need. :-)

Reply:
Wow, that's cool! Thanks for taking the time to explain it.